internal/inspect/policy.go

```diff
@@ -0,0 +1,289 @@
+		sortedLocalVariables := make([]*responsev1.InspectPoliciesResponse_Variable, 0, len(localVariables))
+		for _, lv := range localVariables {
+			sortedLocalVariables = append(sortedLocalVariables, lv)
+		}
+		sort.Slice(sortedLocalVariables, func(i, j int) bool {
+			return sortedLocalVariables[i].Name < sortedLocalVariables[j].Name
+		})
+
+		if len(localVariables) > 0 {
+			pol.results[policyID] = &responsev1.InspectPoliciesResponse_Result{
+				Variables: sortedLocalVariables,
+			}
+		}
+
+		return nil
+	}
+
+	referencedVariables, err := pol.inspectConditions(p)
+	if err != nil {
+		return fmt.Errorf("failed to inspect conditions of the policy %s: %w", policyID, err)
+	}
+	pol.imports[policyID] = pol.listImports(p)
+
+	for referenced := range referencedVariables {
+		if v, ok := localVariables[referenced]; ok {
+			v.Used = true
+		} else {
+			if toResolve, exists := pol.toResolve[policyID]; !exists {
+				pol.toResolve[policyID] = map[string]bool{referenced: false}
+			} else {
+				toResolve[referenced] = false
+			}
+		}
+	}
+
+	variables := make([]*responsev1.InspectPoliciesResponse_Variable, 0, len(localVariables))
+	for _, lv := range localVariables {
+		variables = append(variables, lv)
+	}
+
+	// sort variables if there is nothing to resolve since we are not going to modify variables in the future.
+	if len(pol.toResolve[policyID]) > 0 {
+		sort.Slice(variables, func(i, j int) bool {
+			return variables[i].Name < variables[j].Name
+		})
+	}
+
+	a := policy.ListActions(p)
+	sort.Strings(a)
+	pol.results[policyID] = &responsev1.InspectPoliciesResponse_Result{
+		Actions:   a,
+		Variables: variables,
+	}
+
+	return nil
+}
+
+type loadPolicyFn func(ctx context.Context, policyKey ...string) ([]*policy.Wrapper, error)
+
+// Results returns the final inspection results.
+func (pol *Policy) Results(ctx context.Context, loadPolicy loadPolicyFn) (map[string]*responsev1.InspectPoliciesResponse_Result, error) {
+	for policyID, variables := range pol.toResolve {
+		importedPolicies, ok := pol.imports[policyID]
+		var missingPolicies []string
+		if ok {
+			for _, importedPolicyID := range importedPolicies {
+				if importedResult, ok := pol.results[importedPolicyID]; ok {
+					for _, importedVariable := range importedResult.Variables {
+						if _, ok := variables[importedVariable.Name]; ok {
+							pol.results[policyID].Variables = append(pol.results[policyID].Variables, &responsev1.InspectPoliciesResponse_Variable{
+								Name:   importedVariable.Name,
+								Value:  importedVariable.Value,
+								Kind:   responsev1.InspectPoliciesResponse_Variable_KIND_IMPORTED,
+								Source: importedPolicyID,
+								Used:   true,
+							})
+							variables[importedVariable.Name] = true
+						}
+					}
+				} else {
+					missingPolicies = append(missingPolicies, importedPolicyID)
+				}
+			}
+		}
+
+		if loadPolicy != nil {
+			for _, importedPolicyID := range missingPolicies {
+				importedPolicy, err := loadPolicy(ctx, importedPolicyID)
```

Comment:
Wouldn't it be more efficient to load all the missing policies at once? Saves some roundtrips to the backing store if it's a remote one like S3 or a database.